### PR TITLE
A bunch of changes

### PR DESCRIPTION
**Pull request by eyalroz**
A bunch of changes I've had in mind, including bug fixes / alternative fixes. Please read the individual commit messages.

Also, I think I might have a couple more suggestions when all of this is done.

**Note:** The overall length of `dylib.hpp` is reduced by some 25%.

### CODE REVIEW

File: include/dylib.hpp [hunk 1]
```diff
@@ -93,49 +56,80 @@ class dylib {
...
+        if (!name) {
+            throw std::invalid_argument("null symbol name");
+        }
+        if (!m_handle) {
+            throw std::logic_error("The dynamic (shared) library handle is null. This should not be possible");
+        }
+
+        auto symbol = _get_symbol(m_handle, name);
+        if (symbol == nullptr) {
+            auto msg = std::string("Failed locating symbol ") + name + " in a dynamic (shared) library: " + dlerror();
```
Comment (martin-olivier) [inline include/dylib.hpp:126]: - `dlerror` is unix only
- `dlerror` can return `nullptr`

File: include/dylib.hpp [hunk 2]
```diff
@@ -215,77 +197,41 @@ class dylib {
...
+        std::string description;
         auto lang = MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US);
-        const DWORD len = FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM, nullptr, error_code, lang, msg, 512, nullptr);
-        if (len > 0)
-            return msg;
-        return nullptr;
-    }
+        const DWORD length =
+            FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM, nullptr, error_code, lang, description, buf_size, nullptr);
+        return (length > 0) ? description : nullptr;
```
Comment (martin-olivier) [inline include/dylib.hpp:229]: build an `std::string` with `nullptr` will result in a crash on linux

Comment (eyalroz) [reply include/dylib.hpp:229]: Hmm, that's right... I need to replace that with an empty string.

**Comment by eyalroz** [comment]
Made the changes you requested.

**Review by martin-olivier — APPROVED** [review]
LGTM!

**Pull request merged**